pallets/gear-builtin-actor/src/lib.rs

```diff
@@ -0,0 +1,403 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! # Gear Builtin Actors Pallet
+//!
+//! The Builtn Actors pallet provides a registry of the builtin actors available in the Runtime.
+//!
+//! - [`Config`]
+//!
+//! ## Overview
+//!
+//! The pallet implements the `pallet_gear::BuiltinRouter` allowing to restore builtin actors
+//! claimed `BuiltinId`'s based on their corresponding `ProgramId` address.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::items_after_test_module)]
+
+extern crate alloc;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+mod migrations;
+
+pub mod weights;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+use alloc::string::ToString;
+use core_processor::common::{DispatchOutcome, JournalNote};
+use gear_core::{
+    ids::{BuiltinId, MessageId, ProgramId},
+    message::{ReplyMessage, ReplyPacket, StoredDispatch},
+};
+use gear_core_errors::SimpleExecutionError;
+use impl_trait_for_tuples::impl_for_tuples;
+pub use pallet_gear::BuiltinRouter;
+use parity_scale_codec::{Decode, Encode};
+use sp_io::hashing::blake2_256;
+use sp_runtime::traits::{TrailingZeroInput, Zero};
+use sp_std::prelude::*;
+pub use weights::WeightInfo;
+
+pub use pallet::*;
+
+#[allow(dead_code)]
+const LOG_TARGET: &str = "gear::builtin_actor";
+
+pub trait Dispatchable {
```

Comment:
> I think it's no need to introduce this trait as the message format will be dictated by this pallet anyway

We don't force a specific message type anywhere in the code. Everything a builtin actor should know about a message is that it has a source, a destination (supposedly, this actor's `BuiltinId`) and a payload convertible to bytes. I personally don't see a problem abstracting away the concrete message type.

Otherwise, your argument can be applied to a thousand different places in Substrate code (like, for instance, every pallet being the only implementation of its `Config` trait etc.).

Reply:
> I think it's no need to introduce this trait as the message format will be dictated by this pallet anyway

We don't force a specific message type anywhere in the code. Everything a builtin actor should know about a message is that it has a source, a destination (supposedly, this actor's `BuiltinId`) and a payload convertible to bytes. I personally don't see a problem abstracting away the concrete message type.

Otherwise, your argument can be applied to a thousand different places in Substrate code (like, for instance, every pallet being the only implementation of its `Config` trait etc.).

Reply:
then ok, if the purpose is to abstract message format from built-in actors themselves. But you have `type BuiltinActor: BuiltinActor<SimpleBuiltinMessage, u64>;` in the pallet `Config`